compiler/qsc_eval/src/lib.rs

```diff
@@ -1082,6 +1083,67 @@ impl State {
         self.set_val_register(Value::Range(val::Range { start, step, end }.into()));
     }
 
+    fn eval_struct(&mut self, copy: Option<ExprId>, fields: &[FieldAssign]) {
+        // Extract a flat list of field indexes.
+        let field_indexes = fields
+            .iter()
+            .map(|f| match &f.field {
+                Field::Path(path) => match path.indices.as_slice() {
+                    &[i] => i,
+                    _ => panic!("field path for struct should have a single index"),
+                },
+                _ => panic!("invalid field for struct"),
+            })
+            .collect::<Vec<_>>();
+
+        if copy.is_some() {
+            // Get the field values and the copy struct value.
+            let field_vals = self.pop_vals(fields.len() + 1);
+            let (copy, field_vals) = field_vals.split_first().expect("copy value is expected");
+
+            // Make a clone of the copy struct value.
+            let mut strct = match copy {
+                Value::Tuple(vals) => {
+                    let mut clone_vals = vec![];
+                    for val in vals.iter() {
+                        clone_vals.push(val.clone());
+                    }
+                    clone_vals
+                }
+                _ => panic!("tuple value is expected for struct copy value"),
+            };
```

Comment:
I think there is an easier way to do this with the helper utilities on `Value` and `Rc`:

```suggestion
            let mut strct = copy.unwrap_tuple().to_vec();
```

You don't need to clone each value manually here because the `.to_vec()` performs the clone for you.

Reply:
Nice, that is much cleaner! I'll have to do a `copy.clone()` in there though because the copy is a shared ref, but still much better.